File: 04-estructuras-datos/11-grafos-representacion/java/implementaciones/GrafoListaAdyacencia.java
import java.util.*;

public class GrafoListaAdyacencia {
    private final Map<String, Set<String>> adyacencias = new HashMap<>();

    public void agregarVertice(String vertice) {
        adyacencias.computeIfAbsent(vertice, k -> new LinkedHashSet<>());
    }

    public void conectar(String a, String b) {
        agregarVertice(a); agregarVertice(b);
        // En un grafo no dirigido cada arista se registra en ambos sentidos.
        adyacencias.get(a).add(b);
        adyacencias.get(b).add(a);
    }

    public Set<String> vecinos(String vertice) {
        return Collections.unmodifiableSet(adyacencias.getOrDefault(vertice, Set.of()));
    }
}

File: 04-estructuras-datos/11-grafos-representacion/scala/implementaciones/GrafoListaAdyacencia.scala
import scala.collection.mutable

final class GrafoListaAdyacencia:
  private val adyacencias = mutable.Map.empty[String, mutable.Set[String]]

  def conectar(a: String, b: String): Unit =
    // Un grafo no dirigido registra la conexión en ambos sentidos.
    adyacencias.getOrElseUpdate(a, mutable.LinkedHashSet.empty) += b
    adyacencias.getOrElseUpdate(b, mutable.LinkedHashSet.empty) += a

  def vecinos(v: String): Set[String] = adyacencias.get(v).map(_.toSet).getOrElse(Set.empty)
